rationalize specialize_int_float

avikchaudhuri opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95099

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

avikchaudhuri committed:
rationalize specialize_int_float



Differential Revision: [D43408128](https://our.internmc.facebook.com/intern/diff/D43408128/)

[ghstack-poisoned]

ezyang commented on torch/_dynamo/config.py:
```diff
@@ -60,6 +57,9 @@
 dynamic_shapes = os.environ.get("TORCHDYNAMO_DYNAMIC_SHAPES") == "1"
 
+# specialize int/float by default unless dynamic_shapes is on
+specialize_int_float = not dynamic_shapes
```
this doesn't do what you think it does

avikchaudhuri replied:
OK, reverted. Other parts of the PR should be sufficient.

avikchaudhuri committed:
Update on "rationalize specialize_int_float"


Differential Revision: [D43408128](https://our.internmc.facebook.com/intern/diff/D43408128/)

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/_dynamo/eval_frame.py:
```diff
@@ -117,11 +117,11 @@ def innermost_fn(fn):
         yield
         return
-    with config.patch(dynamic_shapes=True, specialize_int_float=False):
+    with config.patch(dynamic_shapes=True, specialize_int_float=specialize_int_float):
```
just read off the config? 
```suggestion
    with config.patch(dynamic_shapes=True, specialize_int_float=config.specialize_int_float):
```

voznesenskym replied:
Setting dynamic should not set it to False, or True - just remove this? The idea is, I think, that dynamic doesn't touch it. Instead, its up to other regimes (eager vs export) to set the flag.

voznesenskym commented:
Hmmm, if I set the config to specialize_int_float one way, but then export, and it silently changes it, thats (a) really hard to patch (b) ignores user config.

Can we instead start by just:

1) Removing the implication that setting dynamic disables it?
2) Asserting a specific config state for this flag if we are in export?

voznesenskym reviewed (changes requested):
Back to you, good first pass :)

ezyang commented:
related https://github.com/pytorch/pytorch/issues/94640

avikchaudhuri committed:
Update on "rationalize specialize_int_float"


Differential Revision: [D43408128](https://our.internmc.facebook.com/intern/diff/D43408128/)

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -590,6 +590,7 @@ def result_capturing_wrapper(*graph_inputs):
 
     remove_from_cache(f)
     with patch(f"{__name__}.most_recent_backend", None):
+        config.specialize_int_float = True
```
this looks bad, why not use patch

ezyang reviewed (approved):
you need to update the xpass. There's also some unrelated refactoring (`enable_dynamic` to `set_dynamic`) which I am not exactly sure the motivation of. Also please use patch to update config for export.

ezyang commented:
It seems like we should delete `test_unspec.py` now

avikchaudhuri committed:
Update on "rationalize specialize_int_float"


Differential Revision: [D43408128](https://our.internmc.facebook.com/intern/diff/D43408128/)

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri commented:
Addressed latest round of comments on code. Also removed expected failure of test_builtin_getitem, but not sure about other failing tests.

voznesenskym reviewed (approved):
A little different than what I requested, but looks fine. I dislike silent swapping of configs and would have preferred an assert, but this works :)

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -590,14 +590,15 @@ def result_capturing_wrapper(*graph_inputs):
-        )(f)
-        # TODO(voz): We may have instances of `f` that mutate inputs, we should track sideffects and reject.
-        result_traced = opt_f(*args, **kwargs)
+        with config.patch(specialize_int_float=True):
```
you can put the two context managers together: `with patch(...), config.patch(...):`

avikchaudhuri committed:
Update on "rationalize specialize_int_float"


Differential Revision: [D43408128](https://our.internmc.facebook.com/intern/diff/D43408128/)

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri committed:
Update on "rationalize specialize_int_float"


Differential Revision: [D43408128](https://our.internmc.facebook.com/intern/diff/D43408128/)

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri committed:
Update on "rationalize specialize_int_float"


Differential Revision: [D43408128](https://our.internmc.facebook.com/intern/diff/D43408128/)

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on test/dynamo/test_misc.py:
```diff
@@ -2114,14 +2114,14 @@ def forward(self, x, ref_id):
         opt_m = torch._dynamo.optimize(cnts, nopython=True)(m)
         opt_m(data, correct_ref_id)
-        self.assertEqual(cnts.op_count, 2)
+        self.assertEqual(cnts.op_count, 3)
```
you can't do this, the test is run with both dynamic shapes enabled and not. Use ifdyn to toggle between the op counts

ezyang commented on test/dynamo/test_unspec.py:
```diff
@@ -172,8 +172,7 @@ def fn(x):
 
-    # TypeError: zeros(): argument 'size' (position 1) must be tuple of SymInts, not FakeTensor
-    @unittest.expectedFailure
+    @torch._dynamo.config.patch("specialize_int_float", True)
```
This diff makes no sense, this is `test_unspec.py`, it's specifically about testing `specialize_int_float=False`

ezyang committed:
Update on "rationalize specialize_int_float"


Differential Revision: [D43408128](https://our.internmc.facebook.com/intern/diff/D43408128/)

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang committed:
Update on "rationalize specialize_int_float"


Differential Revision: [D43408128](https://our.internmc.facebook.com/intern/diff/D43408128/)

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang committed:
Update on "rationalize specialize_int_float"


Differential Revision: [D43408128](https://our.internmc.facebook.com/intern/diff/D43408128/)

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang committed:
Update on "rationalize specialize_int_float"


Differential Revision: [D43408128](https://our.internmc.facebook.com/intern/diff/D43408128/)

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on test/dynamo/test_dynamic_shapes.py:
```diff
@@ -114,6 +114,16 @@ def make_dynamic_cls(cls, assume_static_by_default):
+
+unittest.expectedFailure(
+    DynamicShapesMiscTests.test_slice_input_dynamic_shapes
+    # NotImplementedError: SymNodeVariable() is not a constant
```
I added this xfail. This is fixable but kind of annoying.

What we would like to happen here is that we force a guard on the int when it slices into a Python list. But the usage of the int happens deep inside a slice, so it is not obvious at the call site that this is possible.  I also tried making `as_python_constant` work universally for SymNodeVariable, but this causes a ton of fallout (as code now assumes that it can exercise this path.) Probably the right thing to do is add another method for this sort of thing.

ezyang committed:
Update on "rationalize specialize_int_float"


Differential Revision: [D43408128](https://our.internmc.facebook.com/intern/diff/D43408128/)

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang committed:
Update on "rationalize specialize_int_float"


Differential Revision: [D43408128](https://our.internmc.facebook.com/intern/diff/D43408128/)

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang committed:
Update on "rationalize specialize_int_float"


Differential Revision: [D43408128](https://our.internmc.facebook.com/intern/diff/D43408128/)

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge -f "unrelated failure"

ezyang commented:
@pytorchbot merge -f "unrelated failure"

ezyang committed:
Update on "rationalize specialize_int_float"


cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
since @tugsbayasgalan landed an unblock, I'm just folding this into https://github.com/pytorch/pytorch/pull/95621
